File: real-time-risk/python/trade_ledger/event_store.py
"""An **app-local** DynamoDB-backed event store — the single biggest parity gap this port confirms.

**Why this lives in the app, not the framework.** .NET ships ``Benzene.EventSourcing`` /
``Benzene.EventSourcing.DynamoDb`` — an ``IEventStore`` with optimistic-concurrency append and a
DynamoDB store the Trade Ledger simply *inherits*. benzene-python has **no event-sourcing package at
all** (confirmed in ``real-time-risk/PARITY-FINDINGS.md`` §3.1 and its capability matrix §1: "Event
sourcing / DynamoDB-backed event store — ✅ .NET, ❌ Python"). So the Trade Ledger has to own the
~100 lines of "conditional-write append + query read" itself, re-implemented against the DynamoDB SDK,
exactly the per-language duplication the parity audit calls out. This class is that implementation.

**Wire/item-shape parity.** The item shape is byte-for-byte the shape .NET's ``DynamoDbEventStore``
writes and the shared Terraform table (`real-time-risk/deploy/terraform/dynamodb.tf`) provisions, so a
DynamoDB Stream consumer (the Risk Read Models projector) sees identical records regardless of which
language wrote them:

======================  ====  ==============================================================
attribute               type  meaning
======================  ====  ==============================================================
``pk``                  S     stream id (the book id) — the partition key
``version``             N     1-based per-stream sequence number — the sort key
``eventType``           S     the event discriminator (e.g. ``TradeBooked``)
``payload``             S     the event body as a JSON string
``timestamp``           S     ISO-8601 UTC instant the event was appended
======================  ====  ==============================================================

**Optimistic concurrency.** ``append`` issues a single ``TransactWriteItems`` with one ``Put`` per new
event, each guarded by ``ConditionExpression="attribute_not_exists(pk)"``. On a composite-key table
that condition is evaluated against the *exact* ``(pk, version)`` item being written, so it succeeds
only when that version slot is still free — precisely .NET's ``attribute_not_exists(#pk)`` guard. If
any Put's condition fails (a concurrent writer already took the version) DynamoDB cancels the whole
transaction; we surface that as :class:`EventStoreConcurrencyError`.

``boto3`` is synchronous, so each SDK call is dispatched through ``asyncio.to_thread`` to keep the
handler's event loop free — the store's public surface is ``async`` to match the handler contract.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


@dataclass(frozen=True)
class EventEnvelope:
    """An event to append: its type discriminator and its already-serialized JSON payload.

    Mirrors .NET ``Benzene.EventSourcing.EventEnvelope(EventType, Payload)``.
    """

    event_type: str
    payload: str


@dataclass(frozen=True)
class StoredEvent:
    """An event read back from a stream: envelope fields plus the version it landed at.

    Mirrors .NET's ``StoredEvent`` (``EventType`` / ``Payload`` / ``Version`` / ``Timestamp``).
    """

    version: int
    event_type: str
    payload: str
    timestamp: str


class EventStoreConcurrencyError(Exception):
    """Raised when an :meth:`DynamoDbEventStore.append` lost an optimistic-concurrency race.

    Mirrors .NET ``EventStoreConcurrencyException``: the expected version no longer matches the
    stream's head, so the caller must re-read and retry.
    """


class DynamoDbEventStore:
    """A per-stream, append-only event log stored one item per event in a DynamoDB table."""

    def __init__(self, client: Any, table_name: str) -> None:
        self._client = client
        self._table_name = table_name

    async def read(self, stream_id: str) -> list[StoredEvent]:
        """Read a stream's full history in version order (``Query pk=:pk`` ascending).

        Mirrors .NET ``DynamoDbEventStore.ReadAsync``. An unknown stream returns ``[]``.
        """
        return await asyncio.to_thread(self._read_sync, stream_id)

    def _read_sync(self, stream_id: str) -> list[StoredEvent]:
        events: list[StoredEvent] = []
        start_key: dict[str, Any] | None = None
        while True:
            kwargs: dict[str, Any] = {
                "TableName": self._table_name,
                "KeyConditionExpression": "pk = :pk",
                "ExpressionAttributeValues": {":pk": {"S": stream_id}},
                "ScanIndexForward": True,  # ascending by version (the range key)
            }
            if start_key is not None:
                kwargs["ExclusiveStartKey"] = start_key
            response = self._client.query(**kwargs)
            for item in response.get("Items", []):
                events.append(
                    StoredEvent(
                        version=int(item["version"]["N"]),
                        event_type=item.get("eventType", {}).get("S", ""),
                        payload=item.get("payload", {}).get("S", ""),
                        timestamp=item.get("timestamp", {}).get("S", ""),
                    )
                )
            start_key = response.get("LastEvaluatedKey")
            if not start_key:
                break
        return events

    async def append(
        self, stream_id: str, expected_version: int, events: list[EventEnvelope]
    ) -> int:
        """Append events after ``expected_version``; return the stream's new highest version.

        Mirrors .NET ``DynamoDbEventStore.AppendAsync``. Each event is written at
        ``expected_version + n`` (1-based) inside one ``TransactWriteItems``, every Put guarded by
        ``attribute_not_exists(pk)`` so a concurrent writer that already claimed a version cancels the
        transaction — surfaced as :class:`EventStoreConcurrencyError`.
        """
        if not events:
            return expected_version
        return await asyncio.to_thread(self._append_sync, stream_id, expected_version, events)

    def _append_sync(
        self, stream_id: str, expected_version: int, events: list[EventEnvelope]
    ) -> int:
        from botocore.exceptions import ClientError

        now = datetime.now(timezone.utc).isoformat()
        transact_items = []
        new_version = expected_version
        for envelope in events:
            new_version += 1
            transact_items.append(
                {
                    "Put": {
                        "TableName": self._table_name,
                        "Item": {
                            "pk": {"S": stream_id},
                            "version": {"N": str(new_version)},
                            "eventType": {"S": envelope.event_type},
                            "payload": {"S": envelope.payload},
                            "timestamp": {"S": now},
                        },
                        # Optimistic concurrency: this (pk, version) slot must still be free.
                        "ConditionExpression": "attribute_not_exists(pk)",
                    }
                }
            )

        try:
            self._client.transact_write_items(TransactItems=transact_items)
        except ClientError as ex:
            code = ex.response.get("Error", {}).get("Code", "")
            # A cancelled transaction (a version slot was taken) is a lost concurrency race, not an
            # infrastructure fault. TransactionCanceledException is the batch-level code; the
            # per-item ConditionalCheckFailed reasons are inside CancellationReasons.
            if code in ("TransactionCanceledException", "ConditionalCheckFailedException"):
                raise EventStoreConcurrencyError(
                    f"Concurrent append to stream {stream_id!r} at expected version "
                    f"{expected_version} (a newer version already exists)."
                ) from ex
            raise
        return new_version

File: real-time-risk/go/eventstore/eventstore.go
// Package eventstore is an app-local, DynamoDB-backed event store for the Trade Ledger.
//
// It exists ONLY because benzene-go has no event-sourcing package. In .NET this is framework code
// - Benzene.EventSourcing (IEventStore, EventEnvelope, optimistic-concurrency AppendAsync) plus
// Benzene.EventSourcing.DynamoDb (DynamoDbEventStore, the pk/version/eventType/payload/timestamp
// item shape) - and the Trade Ledger simply inherits it. No non-.NET port has lifted event
// sourcing into the framework yet, so every other language re-implements this same ~150 lines of
// "conditional-write append + query read" by hand. This package is that hand-rolled equivalent
// for Go, documented as app-local per PARITY-FINDINGS.md §3.1 ("Event sourcing is a .NET-only
// capability... this repo's ports implement the store as app-local code").
//
// The DynamoDB item shape is byte-for-byte the shape the .NET DynamoDbEventStore writes and the
// shared Terraform table (real-time-risk/deploy/terraform/dynamodb.tf) provisions, so the store is
// wire-compatible across languages against one physical table:
//
//	pk        (S, partition) = stream / book id
//	version   (N, sort)      = 1-based sequence within the stream
//	eventType (S)            = the event-type discriminator (e.g. "TradeBooked")
//	payload   (S)            = the event body as a JSON string
//	timestamp (S)            = RFC3339 append time
package eventstore

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
)

// Event is one event to append: its type discriminator and its already-serialized JSON payload -
// the Go counterpart of .NET's EventEnvelope.
type Event struct {
	EventType string
	Payload   string
}

// StoredEvent is one event as read back from the store, with its assigned stream position and
// append time.
type StoredEvent struct {
	StreamID  string
	Version   int64
	EventType string
	Payload   string
	Timestamp time.Time
}

// ErrConcurrency is returned by Append when the optimistic-concurrency condition fails: some
// other writer already appended at one of the versions this call tried to claim (the ledger's
// TransactWriteItems was cancelled). The .NET analogue is EventStoreConcurrencyException. A caller
// that wants last-writer-wins retries would re-read the stream and retry the append.
var ErrConcurrency = errors.New("eventstore: optimistic concurrency conflict (stream advanced concurrently)")

// Store is the event-sourcing read/append surface the Trade Ledger depends on. Declaring it as an
// interface (rather than the concrete DynamoStore) is what lets the handler's tests run against an
// in-memory fake with no DynamoDB, per the task's "event-store integration against real DynamoDB
// is out of scope for local unit tests" note.
type Store interface {
	// Read returns every event on streamID in ascending version order (empty for an unknown stream).
	Read(ctx context.Context, streamID string) ([]StoredEvent, error)
	// Append writes events at versions expectedVersion+1, +2, ... atomically, failing with
	// ErrConcurrency if any of those versions already exists. It returns the new highest version.
	Append(ctx context.Context, streamID string, expectedVersion int64, events []Event) (newVersion int64, err error)
}

// DynamoStore is the DynamoDB-backed Store. It is safe for concurrent use (the AWS SDK client is).
type DynamoStore struct {
	client    *dynamodb.Client
	tableName string
}

// NewDynamoStore builds a DynamoStore over client for tableName.
func NewDynamoStore(client *dynamodb.Client, tableName string) *DynamoStore {
	return &DynamoStore{client: client, tableName: tableName}
}

var _ Store = (*DynamoStore)(nil)

// Read queries `pk = :pk` with ScanIndexForward=true so events come back in ascending version
// order - the stream's history, oldest first.
func (s *DynamoStore) Read(ctx context.Context, streamID string) ([]StoredEvent, error) {
	out, err := s.client.Query(ctx, &dynamodb.QueryInput{
		TableName:              aws.String(s.tableName),
		KeyConditionExpression: aws.String("pk = :pk"),
		ExpressionAttributeValues: map[string]types.AttributeValue{
			":pk": &types.AttributeValueMemberS{Value: streamID},
		},
		ScanIndexForward: aws.Bool(true),
	})
	if err != nil {
		return nil, fmt.Errorf("eventstore: read stream %q: %w", streamID, err)
	}

	events := make([]StoredEvent, 0, len(out.Items))
	for _, item := range out.Items {
		stored, err := itemToStoredEvent(streamID, item)
		if err != nil {
			return nil, err
		}
		events = append(events, stored)
	}
	return events, nil
}

// Append writes each event as its own item at the next version, all in one TransactWriteItems so
// the append is atomic. Every Put carries ConditionExpression `attribute_not_exists(pk)`: on a
// Put keyed by (pk, version) that condition is true only when no item with that exact key exists,
// which is the optimistic-concurrency guard - if a concurrent writer already claimed the version,
// the whole transaction is cancelled and we surface ErrConcurrency. This mirrors the .NET
// DynamoDbEventStore's TransactWriteItems + attribute_not_exists(#pk) append.
func (s *DynamoStore) Append(ctx context.Context, streamID string, expectedVersion int64, events []Event) (int64, error) {
	if len(events) == 0 {
		return expectedVersion, nil
	}

	timestamp := time.Now().UTC().Format(time.RFC3339)
	writes := make([]types.TransactWriteItem, 0, len(events))
	version := expectedVersion
	for _, event := range events {
		version++
		writes = append(writes, types.TransactWriteItem{
			Put: &types.Put{
				TableName: aws.String(s.tableName),
				Item: map[string]types.AttributeValue{
					"pk":        &types.AttributeValueMemberS{Value: streamID},
					"version":   &types.AttributeValueMemberN{Value: strconv.FormatInt(version, 10)},
					"eventType": &types.AttributeValueMemberS{Value: event.EventType},
					"payload":   &types.AttributeValueMemberS{Value: event.Payload},
					"timestamp": &types.AttributeValueMemberS{Value: timestamp},
				},
				// attribute_not_exists(pk) => "no item with this (pk, version) key yet" => optimistic lock.
				ConditionExpression: aws.String("attribute_not_exists(pk)"),
			},
		})
	}

	_, err := s.client.TransactWriteItems(ctx, &dynamodb.TransactWriteItemsInput{TransactItems: writes})
	if err != nil {
		var cancelled *types.TransactionCanceledException
		if errors.As(err, &cancelled) {
			return 0, fmt.Errorf("%w: stream %q at expected version %d: %v", ErrConcurrency, streamID, expectedVersion, err)
		}
		return 0, fmt.Errorf("eventstore: append to stream %q: %w", streamID, err)
	}
	return version, nil
}

// itemToStoredEvent decodes one DynamoDB item into a StoredEvent, tolerating a missing/typeless
// attribute by treating it as its zero value rather than failing the whole read (a malformed row
// is a data problem, not a reason to 500 every query of the stream).
func itemToStoredEvent(streamID string, item map[string]types.AttributeValue) (StoredEvent, error) {
	stored := StoredEvent{StreamID: streamID}

	if v, ok := item["version"].(*types.AttributeValueMemberN); ok {
		version, err := strconv.ParseInt(v.Value, 10, 64)
		if err != nil {
			return StoredEvent{}, fmt.Errorf("eventstore: stream %q has non-numeric version %q: %w", streamID, v.Value, err)
		}
		stored.Version = version
	}
	if v, ok := item["eventType"].(*types.AttributeValueMemberS); ok {
		stored.EventType = v.Value
	}
	if v, ok := item["payload"].(*types.AttributeValueMemberS); ok {
		stored.Payload = v.Value
	}
	if v, ok := item["timestamp"].(*types.AttributeValueMemberS); ok {
		if ts, err := time.Parse(time.RFC3339, v.Value); err == nil {
			stored.Timestamp = ts
		}
	}
	return stored, nil
}
